### ISSUE TITLE
Concurrent thread-safe inference hangs on Jetson Orin 8GB

### ISSUE THREAD

**Issue by iokarkan** [ultralytics/ultralytics#14891]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I have been trying to run my application that performs concurrent inference with a thread pool executor, that works on an RTX 3060, on a Jetson Orin 8GB. 

In my app, I spawn processes, inside which a number of YOLO models are loaded, and within each process concurrent inference with the process's models is performed with ThreadPoolExecutor (which I now understand may be thread-unsafe). Using an RTX 3060, there's no problem perfroming concurrent inference this way with 3 models, and there has been no observed race condition.

However, running the same code (but with small variation in the venv for compatibility, but same ultralytics version), the app hangs when 3 model objects are asked to infer concurrently within a process. 1 or 2 models works fine.

Following the [thread-safe inference docs section](https://docs.ultralytics.com/guides/yolo-thread-safe-inference/#thread-safe-inference) , I tried creating an MRE for this issue, but I even the N=2 case hangs on the Jetson (works fine on the RTX3060). 

In such a case, after the models are loaded in RAM, the CPU usage falls to about 0 and event ctrl+c will not be responsive. Notably, it also hangs whether I spawn a process or directly use the main process. Finally, refactoring to use a ThreadPoolExecutor also hangs.

### yolo checks
OS                  Linux-5.10.104-tegra-aarch64-with-glibc2.29
Environment         Linux
Python              3.8.10
Install             git
RAM                 7.16 GB
CPU                 ARMv8 Processor rev 1 (v8l)
CUDA                11.4

matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 9.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.28.2>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.1.0a0+41361538.nv23.6>=1.8.0
torchvision         ✅ 0.16.1>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0

### Jetson Info
xxxx@ubuntu:~$ jetson_release -v
Software part of jetson-stats 4.2.1 - (c) 2023, Raffaello Bonghi
Model: NVIDIA Orin Nano Developer Kit - Jetpack 5.1.1 [L4T 35.3.1]
NV Power Mode[0]: 15W
Serial Number: [XXX Show with: jetson_release -s XXX]
Hardware:

699-level Part Number: 699-13767-0003-300 M.2
P-Number: p3767-0003
Module: NVIDIA Jetson Orin Nano (8GB ram)
SoC: tegra23x
CUDA Arch BIN: 8.7
Codename: P3768
Platform:

Machine: aarch64
System: Linux
Distribution: Ubuntu 20.04 focal
Release: 5.10.104-tegra
Python: 3.8.10
jtop:

Version: 4.2.1
Service: Active
Libraries:

CUDA: 11.4.315
cuDNN: 8.6.0.166
TensorRT: 8.5.2.2
VPI: 2.2.7
Vulkan: 1.3.204
OpenCV: 4.5.4 - with CUDA: NO

### Minimal Reproducible Example

```python
import multiprocessing as mp
import threading
import numpy as np
from queue import Queue
from ultralytics import YOLO

# Dummy image array
img = np.random.randint(0, 256, (1440, 720, 3), dtype=np.uint8)

def load_and_infer(model_path, image, queue, idx):
    """ Load the YOLO model, perform inference on the provided image, and put the result into the queue. """
    model = YOLO(model_path)
    result = model(image)
    queue.put((idx, result))

def process():
    """ Main process function to perform inference using threads and a queue. """
    N = 2
    model_path = 'yolov8n.pt'
    queue = Queue()
    threads = []

    print(f"Starting the process with {N} models")

    for i in range(N):
        thread = threading.Thread(target=load_and_infer, args=(model_path, img, queue, i))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    results = [None] * N
    while not queue.empty():
        idx, result = queue.get()
        results[idx] = result

    for idx, result in enumerate(results):
        print(f"Result from model {idx + 1}: {result}")

if __name__ == "__main__":
    #p = mp.Process(target=process)
    #p.start()
    #p.join()
    process()
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by iokarkan** [comment]
> What's the goal of running concurrent inference?

To avoid having the inferences be performed serially, one after another, for real time application.

**Comment by Y-T-G** [comment]
And if you want to increase throughput for single stream, then you should use batching, which will work way better than this approach.

**Comment by iokarkan** [comment]
> But loading a model each time for inference is terrible.

Indeed,  I agree for the first part (without actually thoroughly testing), and in the original application I am not re-initializing the model. For this MRE, however, I tried to follow the thread-safe instructions from the docs, and the problem I am stating is still there. Even if you only perform the model() call in the thread target function, the reported issue remains.

> If you want to use multiple inference, you should use Triton Inference Server to host the model and multiprocessing to create multiple cloents.

For the second part, I have not considered this approach, and maybe it has value but seems rather heavy for the requirements of my application. Although it may be off-topic for this issue, it may help if inference is done concurrently on the Triton server, and not serially.

> And if you want to increase throughput for single stream, then you should use batching, which will work way better than this approach.

As mentioned, this is a real time application, and batching is not available.

**Comment by Y-T-G** [comment]
Are you running multiple streams or single stream?

I am not sure how you're expecting the performance to increase by using threading if it's single. If the GPU is already under load, sending another doesn't seem like it would help especially for Jetson. RTX 3060 might have enough juice for extra inference requests, but Jetson isn't that powerful.

**Comment by Y-T-G** [comment]
But if it's multi-stream, then you can use batching. In fact, Triton Inference Supports dynamic batching and you don't even have to batch them explicitly.

**Comment by iokarkan** [comment]
Thank you for comments.

First, let me note that I am doing inference on a single image, with multiple models that are simultaneously loaded on the GPU (and they fit in Jetson RAM). The purpose of threading is to allow the computations to be done concurrently, achieving faster time than having the python code synchronously wait for the first model's computation to return before starting another's. Also, since I am grabbing from a video feed at real-time, I am performing inference as soon as an image is available.

I am not sure what single/multiple-stream is or of the capabilities of a Triton server, so I do not know how to answer.

But let me kindly state that my issue here is basically an inability to reproduce a part of the docs, despite the fact that (as I undestand it) I should be able to on this Jetson device.

**Comment by Y-T-G** [comment]
Well, just because you have more VRAM left, doesn't mean you have more GPU compute left. They're separate things.

> Also, since I am grabbing from a video feed at real-time, I am performing inference as soon as an image is available.

Yes, but depending on your stream's FPS, that could take 33ms (30FPS). If your inference speed is faster than 30FPS, then you are not going to receive the next frame before the inference finishes. And if it is not faster than 30FPS, then it's bottlenecked by GPU speed, and sending a new inference request when the GPU is already under load will not make the inference any faster.

> But let me kindly state that my issue here is basically an inability to reproduce a part of the docs, despite the fact that (as I undestand it) I should be able to on this Jetson device.

The code seems to work on your more powerful device and fails om the Jetson with a much weaker CPU and GPU. I would say that's expected.

**Comment by Y-T-G** [comment]
> Notably, it also hangs whether I spawn a process or directly use the main process. Finally, refactoring to use a ThreadPoolExecutor also hangs.

Does it hang when you perform inference concurrently with the models loaded in separate processes?

**Comment by lakshanthad** [comment]
Here are my findings running the code shared by @iokarkan on NVIDIA Jetson Orin NX 16GB. Is this behavior similar with your RTX machine?

```sh
root@288b7532b0ea:/ultralytics# python3 test.py
Starting the process with 2 models


0: 640x320 (no detections), 236.3ms
Speed: 4.9ms preprocess, 236.3ms inference, 37.3ms postprocess per image at shape (1, 3, 640, 320)
0: 640x320 (no detections), 234.2ms
Speed: 4.4ms preprocess, 234.2ms inference, 38.1ms postprocess per image at shape (1, 3, 640, 320)
Result from model 1: [ultralytics.engine.results.Results object with attributes:

boxes: ultralytics.engine.results.Boxes object
keypoints: None
masks: None
names: {0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: 'airplane', 5: 'bus', 6: 'train', 7: 'truck', 8: 'boat', 9: 'traffic light', 10: 'fire hydrant', 11: 'stop sign', 12: 'parking meter', 13: 'bench', 14: 'bird', 15: 'cat', 16: 'dog', 17: 'horse', 18: 'sheep', 19: 'cow', 20: 'elephant', 21: 'bear', 22: 'zebra', 23: 'giraffe', 24: 'backpack', 25: 'umbrella', 26: 'handbag', 27: 'tie', 28: 'suitcase', 29: 'frisbee', 30: 'skis', 31: 'snowboard', 32: 'sports ball', 33: 'kite', 34: 'baseball bat', 35: 'baseball glove', 36: 'skateboard', 37: 'surfboard', 38: 'tennis racket', 39: 'bottle', 40: 'wine glass', 41: 'cup', 42: 'fork', 43: 'knife', 44: 'spoon', 45: 'bowl', 46: 'banana', 47: 'apple', 48: 'sandwich', 49: 'orange', 50: 'broccoli', 51: 'carrot', 52: 'hot dog', 53: 'pizza', 54: 'donut', 55: 'cake', 56: 'chair', 57: 'couch', 58: 'potted plant', 59: 'bed', 60: 'dining table', 61: 'toilet', 62: 'tv', 63: 'laptop', 64: 'mouse', 65: 'remote', 66: 'keyboard', 67: 'cell phone', 68: 'microwave', 69: 'oven', 70: 'toaster', 71: 'sink', 72: 'refrigerator', 73: 'book', 74: 'clock', 75: 'vase', 76: 'scissors', 77: 'teddy bear', 78: 'hair drier', 79: 'toothbrush'}
obb: None
orig_img: array([[[187, 188,  39],
        [125, 112, 171],
        [207, 199,  40],
        ...,
        [213, 121, 186],
        [167, 194, 243],
        [ 79,  62, 150]],

       [[185, 202,  19],
        [137, 235, 173],
        [234, 237, 171],
        ...,
        [ 85, 185, 116],
        [ 97,  45, 117],
        [ 38,  99, 105]],

       [[ 17,  27, 172],
        [ 53, 179,  24],
        [  5,  11, 227],
        ...,
        [186, 218, 166],
        [192, 252, 226],
        [ 36, 254,  62]],

       ...,

       [[ 14, 197,  53],
        [247, 104, 141],
        [201, 203,  40],
        ...,
        [ 23, 215,  10],
        [102,  64, 154],
        [221, 104, 188]],

       [[210, 198,  76],
        [ 50, 144,  44],
        [ 45,  18, 210],
        ...,
        [ 30,  34,  29],
        [230, 220, 220],
        [180, 161,  60]],

       [[174,  76,  27],
        [145, 116, 121],
        [ 23, 105, 102],
        ...,
        [190, 131,  33],
        [  6,  76, 184],
        [ 80, 186, 214]]], dtype=uint8)
orig_shape: (1440, 720)
path: 'image0.jpg'
probs: None
save_dir: '/ultralytics/runs/detect/predict'
speed: {'preprocess': 4.39906120300293, 'inference': 234.22574996948242, 'postprocess': 38.13457489013672}]
Result from model 2: [ultralytics.engine.results.Results object with attributes:

boxes: ultralytics.engine.results.Boxes object
keypoints: None
masks: None
names: {0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: 'airplane', 5: 'bus', 6: 'train', 7: 'truck', 8: 'boat', 9: 'traffic light', 10: 'fire hydrant', 11: 'stop sign', 12: 'parking meter', 13: 'bench', 14: 'bird', 15: 'cat', 16: 'dog', 17: 'horse', 18: 'sheep', 19: 'cow', 20: 'elephant', 21: 'bear', 22: 'zebra', 23: 'giraffe', 24: 'backpack', 25: 'umbrella', 26: 'handbag', 27: 'tie', 28: 'suitcase', 29: 'frisbee', 30: 'skis', 31: 'snowboard', 32: 'sports ball', 33: 'kite', 34: 'baseball bat', 35: 'baseball glove', 36: 'skateboard', 37: 'surfboard', 38: 'tennis racket', 39: 'bottle', 40: 'wine glass', 41: 'cup', 42: 'fork', 43: 'knife', 44: 'spoon', 45: 'bowl', 46: 'banana', 47: 'apple', 48: 'sandwich', 49: 'orange', 50: 'broccoli', 51: 'carrot', 52: 'hot dog', 53: 'pizza', 54: 'donut', 55: 'cake', 56: 'chair', 57: 'couch', 58: 'potted plant', 59: 'bed', 60: 'dining table', 61: 'toilet', 62: 'tv', 63: 'laptop', 64: 'mouse', 65: 'remote', 66: 'keyboard', 67: 'cell phone', 68: 'microwave', 69: 'oven', 70: 'toaster', 71: 'sink', 72: 'refrigerator', 73: 'book', 74: 'clock', 75: 'vase', 76: 'scissors', 77: 'teddy bear', 78: 'hair drier', 79: 'toothbrush'}
obb: None
orig_img: array([[[187, 188,  39],
        [125, 112, 171],
        [207, 199,  40],
        ...,
        [213, 121, 186],
        [167, 194, 243],
        [ 79,  62, 150]],

       [[185, 202,  19],
        [137, 235, 173],
        [234, 237, 171],
        ...,
        [ 85, 185, 116],
        [ 97,  45, 117],
        [ 38,  99, 105]],

       [[ 17,  27, 172],
        [ 53, 179,  24],
        [  5,  11, 227],
        ...,
        [186, 218, 166],
        [192, 252, 226],
        [ 36, 254,  62]],

       ...,

       [[ 14, 197,  53],
        [247, 104, 141],
        [201, 203,  40],
        ...,
        [ 23, 215,  10],
        [102,  64, 154],
        [221, 104, 188]],

       [[210, 198,  76],
        [ 50, 144,  44],
        [ 45,  18, 210],
        ...,
        [ 30,  34,  29],
        [230, 220, 220],
        [180, 161,  60]],

       [[174,  76,  27],
        [145, 116, 121],
        [ 23, 105, 102],
        ...,
        [190, 131,  33],
        [  6,  76, 184],
        [ 80, 186, 214]]], dtype=uint8)
orig_shape: (1440, 720)
path: 'image0.jpg'
probs: None
save_dir: '/ultralytics/runs/detect/predict'
speed: {'preprocess': 4.906654357910156, 'inference': 236.2804412841797, 'postprocess': 37.30630874633789}]
```

**Comment by lakshanthad** [comment]
I just ran the example code provided by @iokarkan on Jetson Orin NX 16GB. Are you getting similar results on your RTX machine?

```sh
root@288b7532b0ea:/ultralytics# python3 test.py
Starting the process with 2 models


0: 640x320 (no detections), 236.3ms
Speed: 4.9ms preprocess, 236.3ms inference, 37.3ms postprocess per image at shape (1, 3, 640, 320)
0: 640x320 (no detections), 234.2ms
Speed: 4.4ms preprocess, 234.2ms inference, 38.1ms postprocess per image at shape (1, 3, 640, 320)
Result from model 1: [ultralytics.engine.results.Results object with attributes:

boxes: ultralytics.engine.results.Boxes object
keypoints: None
masks: None
names: {0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: 'airplane', 5: 'bus', 6: 'train', 7: 'truck', 8: 'boat', 9: 'traffic light', 10: 'fire hydrant', 11: 'stop sign', 12: 'parking meter', 13: 'bench', 14: 'bird', 15: 'cat', 16: 'dog', 17: 'horse', 18: 'sheep', 19: 'cow', 20: 'elephant', 21: 'bear', 22: 'zebra', 23: 'giraffe', 24: 'backpack', 25: 'umbrella', 26: 'handbag', 27: 'tie', 28: 'suitcase', 29: 'frisbee', 30: 'skis', 31: 'snowboard', 32: 'sports ball', 33: 'kite', 34: 'baseball bat', 35: 'baseball glove', 36: 'skateboard', 37: 'surfboard', 38: 'tennis racket', 39: 'bottle', 40: 'wine glass', 41: 'cup', 42: 'fork', 43: 'knife', 44: 'spoon', 45: 'bowl', 46: 'banana', 47: 'apple', 48: 'sandwich', 49: 'orange', 50: 'broccoli', 51: 'carrot', 52: 'hot dog', 53: 'pizza', 54: 'donut', 55: 'cake', 56: 'chair', 57: 'couch', 58: 'potted plant', 59: 'bed', 60: 'dining table', 61: 'toilet', 62: 'tv', 63: 'laptop', 64: 'mouse', 65: 'remote', 66: 'keyboard', 67: 'cell phone', 68: 'microwave', 69: 'oven', 70: 'toaster', 71: 'sink', 72: 'refrigerator', 73: 'book', 74: 'clock', 75: 'vase', 76: 'scissors', 77: 'teddy bear', 78: 'hair drier', 79: 'toothbrush'}
obb: None
orig_img: array([[[187, 188,  39],
        [125, 112, 171],
        [207, 199,  40],
        ...,
        [213, 121, 186],
        [167, 194, 243],
        [ 79,  62, 150]],

       [[185, 202,  19],
        [137, 235, 173],
        [234, 237, 171],
        ...,
        [ 85, 185, 116],
        [ 97,  45, 117],
        [ 38,  99, 105]],

       [[ 17,  27, 172],
        [ 53, 179,  24],
        [  5,  11, 227],
        ...,
        [186, 218, 166],
        [192, 252, 226],
        [ 36, 254,  62]],

       ...,

       [[ 14, 197,  53],
        [247, 104, 141],
        [201, 203,  40],
        ...,
        [ 23, 215,  10],
        [102,  64, 154],
        [221, 104, 188]],

       [[210, 198,  76],
        [ 50, 144,  44],
        [ 45,  18, 210],
        ...,
        [ 30,  34,  29],
        [230, 220, 220],
        [180, 161,  60]],

       [[174,  76,  27],
        [145, 116, 121],
        [ 23, 105, 102],
        ...,
        [190, 131,  33],
        [  6,  76, 184],
        [ 80, 186, 214]]], dtype=uint8)
orig_shape: (1440, 720)
path: 'image0.jpg'
probs: None
save_dir: '/ultralytics/runs/detect/predict'
speed: {'preprocess': 4.39906120300293, 'inference': 234.22574996948242, 'postprocess': 38.13457489013672}]
Result from model 2: [ultralytics.engine.results.Results object with attributes:

boxes: ultralytics.engine.results.Boxes object
keypoints: None
masks: None
names: {0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: 'airplane', 5: 'bus', 6: 'train', 7: 'truck', 8: 'boat', 9: 'traffic light', 10: 'fire hydrant', 11: 'stop sign', 12: 'parking meter', 13: 'bench', 14: 'bird', 15: 'cat', 16: 'dog', 17: 'horse', 18: 'sheep', 19: 'cow', 20: 'elephant', 21: 'bear', 22: 'zebra', 23: 'giraffe', 24: 'backpack', 25: 'umbrella', 26: 'handbag', 27: 'tie', 28: 'suitcase', 29: 'frisbee', 30: 'skis', 31: 'snowboard', 32: 'sports ball', 33: 'kite', 34: 'baseball bat', 35: 'baseball glove', 36: 'skateboard', 37: 'surfboard', 38: 'tennis racket', 39: 'bottle', 40: 'wine glass', 41: 'cup', 42: 'fork', 43: 'knife', 44: 'spoon', 45: 'bowl', 46: 'banana', 47: 'apple', 48: 'sandwich', 49: 'orange', 50: 'broccoli', 51: 'carrot', 52: 'hot dog', 53: 'pizza', 54: 'donut', 55: 'cake', 56: 'chair', 57: 'couch', 58: 'potted plant', 59: 'bed', 60: 'dining table', 61: 'toilet', 62: 'tv', 63: 'laptop', 64: 'mouse', 65: 'remote', 66: 'keyboard', 67: 'cell phone', 68: 'microwave', 69: 'oven', 70: 'toaster', 71: 'sink', 72: 'refrigerator', 73: 'book', 74: 'clock', 75: 'vase', 76: 'scissors', 77: 'teddy bear', 78: 'hair drier', 79: 'toothbrush'}
obb: None
orig_img: array([[[187, 188,  39],
        [125, 112, 171],
        [207, 199,  40],
        ...,
        [213, 121, 186],
        [167, 194, 243],
        [ 79,  62, 150]],

       [[185, 202,  19],
        [137, 235, 173],
        [234, 237, 171],
        ...,
        [ 85, 185, 116],
        [ 97,  45, 117],
        [ 38,  99, 105]],

       [[ 17,  27, 172],
        [ 53, 179,  24],
        [  5,  11, 227],
        ...,
        [186, 218, 166],
        [192, 252, 226],
        [ 36, 254,  62]],

       ...,

       [[ 14, 197,  53],
        [247, 104, 141],
        [201, 203,  40],
        ...,
        [ 23, 215,  10],
        [102,  64, 154],
        [221, 104, 188]],

       [[210, 198,  76],
        [ 50, 144,  44],
        [ 45,  18, 210],
        ...,
        [ 30,  34,  29],
        [230, 220, 220],
        [180, 161,  60]],

       [[174,  76,  27],
        [145, 116, 121],
        [ 23, 105, 102],
        ...,
        [190, 131,  33],
        [  6,  76, 184],
        [ 80, 186, 214]]], dtype=uint8)
orig_shape: (1440, 720)
path: 'image0.jpg'
probs: None
save_dir: '/ultralytics/runs/detect/predict'
speed: {'preprocess': 4.906654357910156, 'inference': 236.2804412841797, 'postprocess': 37.30630874633789}]
```

**Comment by iokarkan** [comment]
@lakshanthad Yes, this is the expected output. Can you please verify that you can run this on GPU? After suspecting my venv configuration may be the culprit, I did a fresh venv with only `pip install ultralytics==8.1.27`, and run the code. It runs on CPU, It works and the output is as expected.

**Comment by iokarkan** [comment]
> > Notably, it also hangs whether I spawn a process or directly use the main process. Finally, refactoring to use a ThreadPoolExecutor also hangs.
> 
> Does it hang when you perform inference concurrently with the models loaded in separate processes?

@Y-T-G I did not try this as I would then need to also re-arrange the code to use pipes or multiprocessing Manager queues to communicate inference results back to the main process, so there is some overhead.

**Comment by lakshanthad** [comment]
@iokarkan Could you try on this Docker? This is configured to properly use the GPU.

```sh
t=ultralytics/ultralytics:latest-jetson-jetpack5
sudo docker pull $t && sudo docker run -it --ipc=host --runtime=nvidia $t
```

**Comment by Y-T-G** [comment]
> I did not try this as I would need to re-arrange the code to use pipes or multiprocessing Manager queues to communicate inference results back to the main process, so there is some overhead.

I want to see if the model can run concurrently in different processes. You just have to start two Python scripts that are running inference at the same time. So two inferences running at the same time. If they work, then it's not a device bottleneck. If they don't, then it is a device issue as the two scripts are independent and shouldn't affect the other.

**Comment by iokarkan** [comment]
> > I did not try this as I would need to re-arrange the code to use pipes or multiprocessing Manager queues to communicate inference results back to the main process, so there is some overhead.
> 
> I want to see if the model can run concurrently in different processes. You just have to start two Python scripts that are running inference at the same time. So two inferences running at the same time. If they work, then it's not a device bottleneck. If they don't, then it is a device issue as the two scripts are independent and shouldn't affect the other.

Yes this has been tested, 2 scripts can be run that load 2 models each at the same time, no problem.

**Comment by iokarkan** [comment]
I have managed to be able to run the script (and the app) by downgrading torch and torchvision:
- `torch`: `pip install ../../torch-1.13.0a0+d0d6b1f2.nv22.10-cp38-cp38-linux_aarch64.whl`
- `torchvision`: `export BUILD_VERSION=0.14.1 # pytorch 1.13.1
git clone --branch release/0.14 https://github.com/pytorch/vision torchvision`

So it appears to be some sort of compatibility issue.

**Comment by pderrenger** [comment]
Thank you for the update. It seems like the issue might be related to compatibility between the specific versions of PyTorch and Torchvision you were using. Ensuring compatibility between these libraries is crucial, especially on specialized hardware like the Jetson. If you encounter further issues, please try updating to the latest versions of Ultralytics, PyTorch, and Torchvision to see if the problem persists. If the issue continues, feel free to reach out again.

**Comment by narugo1992** [comment]
For anyone landing here in 2025/2026 with the ONNX flavour of this — the torch/torchvision downgrade fix in this thread is specific to the `.pt` Jetson story. If you instead switch the pipeline to `.onnx` + `onnxruntime-gpu` and share one `InferenceSession` across worker threads, you'll hit a different (and longer-lived) race: ORT's CUDA-EP `Softmax` operator is not thread-safe when its cuDNN handle is shared across a session's threads. The ult export pipeline always emits at least one `Softmax` node for DFL decode in `Detect._inference`, so every v3..v12 size is affected. The trigger threshold scales with GPU: ≥8 worker threads on 1660 Ti, ≥32 on H200, but it's the same bug.

The thread-safe-inference guide's "instantiate a YOLO model inside each thread" recipe avoids the race by giving each thread its own session, but per-thread sessions contend for cuDNN handles + CUDA workspace and we measured throughput dropping past ~4 workers on H200. If shared-session-no-lock is what you want, the race can be removed at the graph level by rewriting each `Softmax` node into the math-equivalent `ReduceMax → Sub → Exp → ReduceSum → Div` subgraph — single-file patcher + reproducer + 33-model coverage matrix: https://github.com/deepghs/yolo-onnx-thread-safe

**Comment by glenn-jocher** [comment]
Thanks for sharing this; since this issue is closed and your report is a different ONNX Runtime/CUDA threading case than the original Jetson `.pt` compatibility issue, please open a new issue for visibility and include a minimal reproducer from our [MRE guide](https://docs.ultralytics.com/help/minimum-reproducible-example/) along with `ultralytics.utils.checks.collect_system_info()`, exact `onnxruntime-gpu`/JetPack/CUDA versions, and any text logs so we can track it separately.

**Comment by narugo1992** [comment]
Thanks — opened #24520 with the MRE, `yolo checks` env, and verbatim ORT crash logs per your list. Trying to keep that one bug-scoped to the ORT-CUDA-EP / `Softmax` concurrency case so it doesn't drift into the Jetson `.pt` story this thread was about.

**Comment by glenn-jocher** [comment]
Thanks for opening `#24520` with the MRE and logs; that’s the right place for the ONNX Runtime/CUDA `Softmax` threading case, so we’ll continue tracking it there and keep this thread scoped to the original Jetson `.pt` compatibility issue.
